packages/sitecore-jss/src/personalize/layout-personalizer.ts

```diff
@@ -34,13 +39,23 @@ export function personalizePlaceholder(
   variantId: string
 ): Array<ComponentRendering | HtmlElementRendering> {
   return components
-    .map((component) =>
-      (component as ComponentRenderingWithExperiences).experiences !== undefined
-        ? (personalizeComponent(component as ComponentRenderingWithExperiences, variantId) as
-            | ComponentRendering
-            | HtmlElementRendering)
-        : component
-    )
+    .map((component) => {
+      const rendering = component as ComponentRendering;
+
+      if ((rendering as ComponentRenderingWithExperiences).experiences !== undefined) {
+        return personalizeComponent(rendering as ComponentRenderingWithExperiences, variantId) as
+          | ComponentRendering
+          | HtmlElementRendering;
+      } else if (rendering.placeholders) {
```

Comment:
Doesn't making this an `else if` mean that if the rendering itself has `.experiences`, then any nested components (in `.placeholders`) wouldn't get personalized? Even though the tests seem to prove otherwise ...

Reply:
@ambrauer `personalizeComponent` function personalizes nested placeholders, because of that we need to do the same in case if `experiences` for wrapper don't exist

Reply:
@illiakovalenko Ah, I see now. I feel like this logic could be simplified (we currently iterate through placeholders in all 3 functions), but this works.